src/Protocol/Document/Server/Proposals/ISemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace OmniSharp.Extensions.LanguageServer.Protocol.Document.Server.Proposals
+{
+    [Obsolete(Constants.Proposal)]
+    [Parallel, Method(DocumentNames.SemanticTokens)]
+    public interface ISemanticTokensHandler : IJsonRpcRequestHandler<SemanticTokensParams, SemanticTokens>,
+        IRegistration<SemanticTokensRegistrationOptions>, ICapability<SemanticTokensCapability>
+    {
+    }
+
+    [Obsolete(Constants.Proposal)]
+    [Parallel, Method(DocumentNames.SemanticTokensEdits)]
+    public interface ISemanticTokensEditsHandler :
+        IJsonRpcRequestHandler<SemanticTokensEditsParams, SemanticTokensOrSemanticTokensEdits>,
+        IRegistration<SemanticTokensRegistrationOptions>, ICapability<SemanticTokensCapability>
+    {
+    }
+
+    [Obsolete(Constants.Proposal)]
+    [Parallel, Method(DocumentNames.SemanticTokensRange)]
+    public interface ISemanticTokensRangeHandler : IJsonRpcRequestHandler<SemanticTokensRangeParams, SemanticTokens>,
+        IRegistration<SemanticTokensRegistrationOptions>, ICapability<SemanticTokensCapability>
+    {
+    }
+
+    [Obsolete(Constants.Proposal)]
+    public abstract class SemanticTokensHandler : ISemanticTokensHandler, ISemanticTokensEditsHandler,
+        ISemanticTokensRangeHandler
+    {
+        private readonly SemanticTokensRegistrationOptions _options;
+
+        public SemanticTokensHandler(SemanticTokensRegistrationOptions registrationOptions)
+        {
+            _options = registrationOptions;
+        }
+
+        public SemanticTokensRegistrationOptions GetRegistrationOptions() => _options;
+
+        public virtual async Task<SemanticTokens> Handle(SemanticTokensParams request,
+            CancellationToken cancellationToken)
+        {
+            var document = await GetSemanticTokensDocument(request, cancellationToken);
+            var builder = document.Create();
+            await Tokenize(builder, request, cancellationToken);
+            return builder.Commit().GetSemanticTokens();
+        }
+
+        public virtual async Task<SemanticTokensOrSemanticTokensEdits> Handle(SemanticTokensEditsParams request,
+            CancellationToken cancellationToken)
+        {
+            var document = await GetSemanticTokensDocument(request, cancellationToken);
+            var builder = document.Edit(request);
+            await Tokenize(builder, request, cancellationToken);
+            return builder.Commit().GetSemanticTokensEdits();
+        }
+
+        // TODO: Is this correct?
```

Comment:
is it? :)

Reply:
I think that was from before I had added the helpers for tokenization.  At this point I think it's good based on my testing, and if it isn't I made it virtual for a reason 😄